File: mobile/lib/features/portfolio/domain/repositories/portfolio_repository.dart
import '../entities/portfolio_item.dart';

/// Abstract repository for portfolio operations.
abstract class PortfolioRepository {
  /// Fetches all portfolio items for a user.
  Future<List<PortfolioItem>> getPortfolioByUser(String userId);

  /// Fetches a single portfolio item by ID.
  Future<PortfolioItem> getPortfolioItem(String id);

  /// Creates a new portfolio item.
  Future<PortfolioItem> createPortfolioItem({
    required String title,
    String? description,
    String? linkUrl,
    required int position,
    List<Map<String, dynamic>>? media,
  });

  /// Updates an existing portfolio item.
  Future<PortfolioItem> updatePortfolioItem(
    String id, {
    String? title,
    String? description,
    String? linkUrl,
    List<Map<String, dynamic>>? media,
  });

  /// Deletes a portfolio item.
  Future<void> deletePortfolioItem(String id);

  /// Reorders portfolio items by providing ordered item IDs.
  Future<void> reorderPortfolio(List<String> itemIds);

  /// Uploads a portfolio image and returns the URL.
  Future<String> uploadPortfolioImage(String filePath);

  /// Uploads a portfolio video and returns the URL.
  Future<String> uploadPortfolioVideo(String filePath);
}

File: backend/internal/port/repository/portfolio_repository.go
package repository

import (
	"context"

	"github.com/google/uuid"

	"marketplace-backend/internal/domain/portfolio"
)

// PortfolioRepository defines persistence operations for portfolio items.
type PortfolioRepository interface {
	// Create inserts a portfolio item with all its media in a transaction.
	Create(ctx context.Context, item *portfolio.PortfolioItem) error

	// GetByID returns a single item with all its media loaded.
	GetByID(ctx context.Context, id uuid.UUID) (*portfolio.PortfolioItem, error)

	// ListByOrganization returns items ordered by position ASC with
	// cursor pagination. Each item includes its media.
	ListByOrganization(ctx context.Context, organizationID uuid.UUID, cursor string, limit int) ([]*portfolio.PortfolioItem, string, error)

	// Update updates title, description, link_url and updated_at.
	Update(ctx context.Context, item *portfolio.PortfolioItem) error

	// Delete removes an item. CASCADE deletes its media.
	Delete(ctx context.Context, id uuid.UUID) error

	// CountByOrganization returns the number of portfolio items for an org.
	CountByOrganization(ctx context.Context, organizationID uuid.UUID) (int, error)

	// ReorderItems batch-updates positions. Index in the slice = new position.
	ReorderItems(ctx context.Context, organizationID uuid.UUID, itemIDs []uuid.UUID) error

	// ReplaceMedia deletes all existing media for an item and inserts new ones.
	ReplaceMedia(ctx context.Context, itemID uuid.UUID, media []*portfolio.PortfolioMedia) error
}
